src/common/jsonb/src/number.rs

```diff
@@ -0,0 +1,262 @@
+        let mut len = bytes.len();
+        assert!(len > 0);
+        len -= 1;
+
+        let ty = bytes[0];
+        match ty {
+            NUMBER_ZERO => Number::UInt64(0),
+            NUMBER_NAN => Number::Float64(f64::NAN),
+            NUMBER_INF => Number::Float64(f64::INFINITY),
+            NUMBER_NEG_INF => Number::Float64(f64::NEG_INFINITY),
+            NUMBER_INT => match len {
+                1 => Number::Int64(i8::from_be_bytes(bytes[1..].try_into().unwrap()) as i64),
+                2 => Number::Int64(i16::from_be_bytes(bytes[1..].try_into().unwrap()) as i64),
+                4 => Number::Int64(i32::from_be_bytes(bytes[1..].try_into().unwrap()) as i64),
+                8 => Number::Int64(i64::from_be_bytes(bytes[1..].try_into().unwrap())),
+                _ => unreachable!(),
+            },
+            NUMBER_UINT => match len {
+                1 => Number::UInt64(u8::from_be_bytes(bytes[1..].try_into().unwrap()) as u64),
+                2 => Number::UInt64(u16::from_be_bytes(bytes[1..].try_into().unwrap()) as u64),
+                4 => Number::UInt64(u32::from_be_bytes(bytes[1..].try_into().unwrap()) as u64),
+                8 => Number::UInt64(u64::from_be_bytes(bytes[1..].try_into().unwrap())),
+                _ => unreachable!(),
+            },
+            NUMBER_FLOAT => Number::Float64(f64::from_be_bytes(bytes[1..].try_into().unwrap())),
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn as_i64(&self) -> Option<i64> {
+        match self {
+            Number::Int64(v) => Some(*v),
+            Number::UInt64(v) => {
+                if *v <= i64::MAX.try_into().unwrap() {
+                    Some(*v as i64)
+                } else {
+                    None
+                }
+            }
+            Number::Float64(_) => None,
+        }
+    }
+
+    pub fn as_u64(&self) -> Option<u64> {
+        match self {
+            Number::Int64(v) => {
+                if *v >= 0 {
+                    Some(*v as u64)
+                } else {
+                    None
+                }
+            }
+            Number::UInt64(v) => Some(*v),
+            Number::Float64(_) => None,
+        }
+    }
+
+    pub fn as_f64(&self) -> Option<f64> {
+        match self {
+            Number::Int64(v) => Some(*v as f64),
+            Number::UInt64(v) => Some(*v as f64),
+            Number::Float64(v) => Some(*v),
+        }
+    }
+}
+
+impl Default for Number {
+    #[inline]
+    fn default() -> Self {
+        Number::UInt64(0)
+    }
+}
+
+impl PartialEq for Number {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl PartialEq<&Number> for Number {
+    #[inline]
+    fn eq(&self, other: &&Number) -> bool {
+        self.eq(*other)
+    }
+}
+
+impl PartialEq<Number> for &Number {
+    #[inline]
+    fn eq(&self, other: &Number) -> bool {
+        (*self).eq(other)
+    }
+}
+
+impl Eq for Number {}
```

Comment:
I’m not sure why the impl is empty? Other codes lgtm.

Reply:
as `Eq` inheritance `PartialEq`, leave it empty is ok.